### PR TITLE
feat(utils): add new functions to convert base64 strings

### LINKED ISSUE #299: New utility functions to convert base64 strings
**Is your feature request related to a problem? Please describe.**

Many projects could benefit from functions to convert buffers to base64 and vice versa, or text to base64 and vice versa.

**Describe the solution you'd like**

`@zk-kit/utils` already provides many functions to convert buffers to other types. Since the `Buffer` library already includes methods to convert base64 strings to other values, it's quite easy to support the conversions mentioned above.

### PR DESCRIPTION
**Pull request by cedoor**
## Description

This PR adds 4 functions to convert base64 strings from buffers and text.

## Related Issue(s)

Closes #299

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] My changes generate no new warnings
-   [x] I have run `yarn style` without getting any errors
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/utils/src/conversions.ts [hunk 1]
```diff
@@ -220,3 +220,49 @@ export function bufferToHexadecimal(value: Buffer | Uint8Array): string {
...
+ * @returns The converted base64 string.
+ */
+export function bufferToBase64(value: Buffer | Uint8Array): string {
+    requireTypes(value, "value", ["Buffer", "Uint8Array"])
+
+    return Buffer.from(value).toString("base64")
+}
+
+/**
+ * Converts a base64 string to a buffer. This function does not check
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:236]: Can you document (and ideally also test in the utest) what happens if it's not?
A quick experiment in Node suggests that what happens is that invalid characters are ignored, rather than an error being thrown.

File: packages/utils/src/conversions.ts [hunk 2]
```diff
@@ -220,3 +220,49 @@ export function bufferToHexadecimal(value: Buffer | Uint8Array): string {
...
+ */
+export function textToBase64(value: string): string {
+    requireString(value, "value")
+
+    return Buffer.from(value, "utf8").toString("base64")
+}
+
+/**
+ * Converts a base64 string to text (utf8). This function does not check
+ * if the input value is a valid base64 string.
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:260]: As above, document what happens to invalid inputs.
In this case there are two different encodings which could go wrong.

File: packages/utils/src/conversions.ts [hunk 3]
```diff
@@ -220,3 +220,49 @@ export function bufferToHexadecimal(value: Buffer | Uint8Array): string {
...
+
+/**
+ * Converts text (utf8) to a base64 string.
+ * @param value The text to convert.
+ * @returns The converted base64 string.
+ */
+export function textToBase64(value: string): string {
+    requireString(value, "value")
+
+    return Buffer.from(value, "utf8").toString("base64")
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:255]: I'm not confident enough in my knowledge of Unicode in JavaScript to be sure this is the right way to deal with non-ASCII text.  I'd suggest some unit tests which exercise this with some non-ASCII text, including multilingual stuff (accents, cyrilic, Chinese) and some emojis, and see that they round-trip correctly through `textToBase64` -> `base64ToText`.

File: packages/utils/src/conversions.ts [hunk 4]
```diff
@@ -220,3 +220,49 @@ export function bufferToHexadecimal(value: Buffer | Uint8Array): string {
 
     return Buffer.from(value).toString("hex")
 }
+
+/**
+ * Converts a buffer to a base64 string. It accepts 'Buffer' or 'Uint8Array'.
+ * @param value The buffer to convert.
+ * @returns The converted base64 string.
+ */
+export function bufferToBase64(value: Buffer | Uint8Array): string {
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:229]: "bytes" instead of "buffer" maybe?  Since the use of the Buffer type is optional for this function and its inverse.

File: packages/utils/tests/conversions.test.ts [hunk 5]
```diff
@@ -170,6 +174,40 @@ describe("Conversions", () => {
         })
     })
 
+    describe("# base64ToBuffer / bufferToBase64", () => {
+        const testBase64 = "SGVsbG8sIFdvcmxkIQ=="
```
Comment (artwyman) [inline packages/utils/tests/conversions.test.ts:178]: Also test some arbitrary byte data which isn't text?  You want this to be able to handle things like null bytes, control characters, etc.

File: packages/utils/tests/conversions.test.ts [hunk 6]
```diff
@@ -170,6 +174,40 @@ describe("Conversions", () => {
...
+
+        it("Should convert a buffer to a base64 string", async () => {
+            const result = bufferToBase64(Buffer.from(testText))
+
+            expect(result).toBe(testBase64)
+        })
+    })
+
+    describe("# base64ToText / textToBase64", () => {
+        const testBase64 = "SGVsbG8sIFdvcmxkIQ=="
```
Comment (artwyman) [inline packages/utils/tests/conversions.test.ts:195]: As I mention above, more test cases using non-ASCII text would be a good idea.

**Review by artwyman — CHANGES_REQUESTED** [review]
Interface and functionality looks good (with the caveat that I may not be sure about the Unicode stuff being right).  My suggestions are about documentation and testing.

File: packages/utils/src/conversions.ts (hunk 1 above)
Comment (cedoor) [reply packages/utils/src/conversions.ts]: I confirm it ignores invalid characters. The documentation also mentions it: https://nodejs.org/api/buffer.html#buffers-and-character-encodings. Checking if a base64 is valid is not trivial and I'd avoid that in this function. I'll add more documentation.

File: packages/utils/src/conversions.ts (hunk 2 above)
Comment (cedoor) [reply packages/utils/src/conversions.ts:260]: In this case `base64ToText` -> `textToBase64` gives a result that is different from the input. So I'll also mention the result may be unexpected.

File: packages/utils/src/conversions.ts (hunk 3 above)
Comment (cedoor) [reply packages/utils/src/conversions.ts:256]: `utf-8` supports all characters in the Unicode standard and it's the preferred encoding for web content. I tested many different characters and they all work 👍🏽

**Comment by cedoor** [comment]
> Interface and functionality looks good (with the caveat that I may not be sure about the Unicode stuff being right). My suggestions are about documentation and testing.

Please, take a look again and let me know if the documentation and tests I added are ok.

File: packages/utils/src/conversions.ts (hunk 1 above)
Comment (artwyman) [reply packages/utils/src/conversions.ts]: I think that's fine so long as the behavior is documented.

A regex can check validity of Base64 pretty reasonably, but that's non-trivial at the level of cost of Base64 conversion, yes.  For PODs I used a regex at the level of key/signature decoding, when I know the specific size and other properties I'm looking for.

File: packages/utils/src/conversions.ts (hunk 4 above)
Comment (artwyman) [reply packages/utils/src/conversions.ts:229]: If you're going to refer to bytes not buffer shouldn't that be reflected in the function name, not just the comments?  I.e. consider renaming the functions to byteToBase64 and base64ToBytes?

File: packages/utils/src/conversions.ts (hunk 3 above)
Comment (artwyman) [reply packages/utils/src/conversions.ts:256]: Agreed, I just wasn't sure what the semantics of JavaScript strings were in that regard.  In some languages, strings separate their internal encoding from external encodings.

One case to consider documenting here is what happens to invalid utf8 sequences in the input.  I presume they'll probably get ignored similarly to the invalid base64 in other cases.

**Review by artwyman — APPROVED** [review]
A few more suggestions in comments, but they're optional so you can decide whether or not to make further changes.  Functionally this seems good.

File: packages/utils/src/conversions.ts (hunk 3 above)
Comment (cedoor) [reply packages/utils/src/conversions.ts:256]: Can you make an example? An input like "��" will be correctly converted.

File: packages/utils/src/conversions.ts (hunk 4 above)
Comment (cedoor) [reply packages/utils/src/conversions.ts:229]: Well, we should update many names in this library in that case. I actually think we can consider Uint8Array to be a type of buffer in this context no?

**Comment by cedoor** [comment]
> A few more suggestions in comments, but they're optional so you can decide whether or not to make further changes. Functionally this seems good.

Thank you 🙏🏽  I'll merge this PR. We can open a new one if necessary.

File: packages/utils/src/conversions.ts (hunk 3 above)
Comment (artwyman) [reply packages/utils/src/conversions.ts:256]: Funny.  I don't know if you really meant to send question-mark charaters or if GitHub + Chrome isn't decoding something.
image

I was thinking of strings which contain non-UTF data.  Not every possible sequence of bytes is valid UTF8.  Look at the encding [here](https://en.wikipedia.org/wiki/UTF-8#Encoding) and you can see if the first byte indicates a multi-byte sequence, then the remaining bytes must have high bits 10, and if they don't it would be an invalid encoding.

It may not be worth worrying about, as bad utf8 is much less of a concern than bad base64.  The structure of the code simply makes me think about any decoding being subject to failure.  I think it's fine that the answer here is "we do what JavaScript does" or "we do what Buffer does".  The main bit of clarification I'd be looking for as a user would be whether the result is an exception, or the result is truncated output.

File: packages/utils/src/conversions.ts (hunk 4 above)
Comment (artwyman) [reply packages/utils/src/conversions.ts:229]: At the code level I think it's the opposite?  Buffer is a type of Uint8Array.

If you mean "buffer" as an english word, then yes, a Uint8Array is a way of representing a "buffer" of bytes, but I'd say that's a confusing definition to rely on when there's a class called Buffer involved.

This feels like bikeshedding, though, and I don't really care enough for it to be worth arguing.

File: packages/utils/src/conversions.ts (hunk 3 above)
Comment (cedoor) [reply packages/utils/src/conversions.ts:256]: I actually sent two question marks, but that's a character that should be correctly supported actually right? :D

File: packages/utils/src/conversions.ts (hunk 4 above)
Comment (cedoor) [reply packages/utils/src/conversions.ts:229]: I think it makes sense, as "Buffer" in the function name may suggest it's a "Buffer" type. I'd keep this small inaccuracy tho since we would need to change many names to keep it consistent and it would be a breaking change.
